Fix lane count calculation - use Scheme structs to avoid re-reading/re-interpreting tags

BudgieInWA opened the pull request:
I am submitting this draft PR to get input on the new approach to `driving_lane_directions`, rust style and `tests.yml`. This PR is aiming towards ~~#69~~ #72.

### Are we happy with this overall structure for `driving_lane_directions`? 
bf5df9d8abdad9b72b4a5b64d558784d6695d1e8 shows the concise version, and dc5de27a3ee539cf25735219d2fdd2c8ffc7e677 shows how useful tag validation warnings can be added (thinking about #101 and how an editor would like to use `tags_to_lanes`).

### Any feedback on my rust style?
I want to fit into this project, more than anything else, but couldn't help trying out `is_some_and` and using lots of `unwrap_or`.

### How do the roundtrip tests work?
I couldn't figure it out just yet. Some guidance in the docs about writing and running `tests.yml` tests would be super awesome. (I figured out `cargo test --feature tests`, which I'll add to the docs if I get there first.)

droogmic commented on data/tests.yml:
```diff
@@ -1125,12 +1125,37 @@
     lanes: '1'
     sidewalk: 'no'
   driving_side: right
+  # I don't think this expected output is correct... -- BudgieInWA
```
feel free to fix the test and then either disable it for all languages or fix it in code.

BudgieInWA replied:
One travel lane with `direction: both` seems sensible to me, I will go with that.

I guess we need to decide how to represent the full continuum of unmarked roads, from narrow alleys where no passing is possible, through "1.5 lane" roads where the shoulder is used for passing, to residential streets with no lane markings wide enough for cars parked on both sides *and* two cars to barely pass between them. Tagging seems all over the place (see the [Key:lanes talk page](https://wiki.openstreetmap.org/wiki/Talk:Key:lanes).

droogmic commented on rust/osm2lanes/src/transform/tags_to_lanes/mod.rs:
```diff
@@ -566,58 +568,75 @@ pub fn tags_to_lanes(
+        .and_then(|num| num.parse::<usize>().ok());
+    let num_both_ways = tagged_both_ways.unwrap_or(0);
+    let tagged_bus_lanes = tags
+        .get("lanes:bus")
```
comment for myself: I think we look at this tag again in busway processing, need to see if this is a code smell or something inherent to the problem.

BudgieInWA replied:
It makes sense to me, to look at it is both places if this function is trying to determine the number of lanes definitively. The `lanes:bus` tag is only being looked at because the way is under tagged, and we want to guess that there are some unrestricted lanes as well. Which would suggest that we should potentially look at `busway` tags here too.

droogmic commented on rust/osm2lanes/src/transform/tags_to_lanes/mod.rs:
```diff
@@ -566,58 +568,75 @@ pub fn tags_to_lanes(
+            (Some(l), None) => (l, 0),
+            (Some(l), Some(f)) => {
+                if l != f {
+                    warn!("lanes tag does not agree with lanes:forward on oneway road")
```
we have a warnings Vec to structurally return this kind of information.
As we are a low level library, we need to give full control to the client on what they want to do with warnings

BudgieInWA replied:
This does seem like the right place to generate most of these warnings, so I will upgrade all of the `warn!`s to instead feed into the `RoadWarnings` vec.

droogmic commented on rust/osm2lanes/src/transform/tags_to_lanes/mod.rs:
```diff
@@ -566,58 +568,75 @@ pub fn tags_to_lanes(
+            (Some(l), None, None) => {
+                // Try to divide the lanes equally.
+                if (l - num_both_ways) % 2 != 0 {
+                    warn!("lanes - lanes:both_ways cannot be evenly distributed");
```
hmmmmmmm
@dabreegster I think you made the decision to round in the favour of forward direction.
Do you think that is consistent enough to not raise a warning?

dabreegster replied:
A warning sounds appropriate. Unless the OSM wiki is very clear what this case means, it's a case of missing / ambiguous data. Good catch!

droogmic replied:
https://wiki.openstreetmap.org/wiki/Key:lanes#Extended_tagging

I agree with the warning, no change needed.

droogmic reviewed (changes requested):
Looks good, thanks for the work so far.

You are working on issue #72

droogmic commented:
> but couldn't help trying out is_some_and and using lots of unwrap_or

Looks pretty idiomatic to me, but then again I am not expert.

droogmic commented:
> I figured out cargo test --feature tests

This was a bug, it should be fixed now.

You should be able to run all the tests with just `cargo test`

The roundtrip works by taking the output (only the output), and checking that you can get back to the same spec by going to tags and back again (round trip).

Why not tags -> spec -> tags?
Well: spec to tags is a one to many mapping, so this shouldn't guarantee a roundtrip.

Documentation is very welcome!

dabreegster commented on rust/osm2lanes/src/lib.rs:
```diff
@@ -4,6 +4,8 @@
 //! WARNING: The output specification and all of this code is just being prototyped. Don't depend
 //! on anything yet.
 
+#![feature(is_some_with)]
```
This would require nightly rust to build, right? Since we're already requiring it for formatting, should we just make use of it everywhere?

(I'd push back a little -- I think it would require all downstream consumers to use nightly too. It's a pity, since there are features like this that're quite handy)

droogmic replied:
why nightly?

dabreegster replied:
``#![feature]` may not be used on the stable release channel`
https://github.com/rust-lang/rust/issues/93050
Until this is stabilized, I think this forces us to use nightly rust

droogmic replied:
it seems features like that are unlikely to ever land in stable, I am actually surprised that one went in:
https://github.com/rust-lang/rust/pull/75298
we can go through the cases where it was used here and replace it with something slightly more verbose :)

BudgieInWA replied:
I agree with staying on stable. I didn't know what `#![feature]` was, but IntelliJ suggested it, so I thought I'd get your input on it. Now I know!

I only used it one time, and can easily work around it.

dabreegster reviewed (commented):
Thanks @droogmic for reviewing, feel free to take the lead on this. I may have some time over the weekend to catch up, but other stuff first...

droogmic commented on rust/osm2lanes/src/transform/tags_to_lanes/mod.rs:
```diff
@@ -566,58 +568,75 @@ pub fn tags_to_lanes(
-    (num_driving_fwd, num_driving_back)
+        match (tagged_lanes, tagged_forward, tagged_backward) {
+            (_, Some(f), Some(b)) => {
+                if tagged_lanes.is_some_and(|&l| l != f + b + num_both_ways) {
+                    warn!("lanes tag does not agree with lanes:forward, lanes:backward and lanes:both_ways");
+                }
```
```suggestion
                if let Some(num_lanes) = tagged_lanes {
                    if num_lanes != f + b + num_both_ways {
                        warn!("lanes tag does not agree with lanes:forward, lanes:backward and lanes:both_ways");
                    }
                }
```

droogmic commented:
Agreed, the content and location of the warnings appears correct.

On Sun, 27 Mar 2022, 07:52 Ben Ritter, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In rust/osm2lanes/src/transform/tags_to_lanes/mod.rs
> <https://github.com/a-b-street/osm2lanes/pull/111#discussion_r835855425>:
>
> > +        .get("lanes:bus")
> +        .and_then(|num| num.parse::<usize>().ok());
> +
> +    if oneway.into() {
> +        // TODO support oneway=-1
> +        if tagged_backward.is_some() {
> +            warn!("lanes:backwards tag on a oneway road")
> +        }
> +        if tagged_both_ways.is_some() {
> +            warn!("lanes:both_ways tag on a oneway road")
> +        }
> +        match (tagged_lanes, tagged_forward) {
> +            (Some(l), None) => (l, 0),
> +            (Some(l), Some(f)) => {
> +                if l != f {
> +                    warn!("lanes tag does not agree with lanes:forward on oneway road")
>
> This does seem like the right place to generate most of these warnings, so
> I will upgrade all of the warn!s to instead feed into the RoadWarnings
> vec.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/a-b-street/osm2lanes/pull/111#discussion_r835855425>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACEL2Z6UD6DUGQ2WZS7ZFEDVB7ZRTANCNFSM5RTT6CFA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

droogmic commented on rust/osm2lanes/src/transform/tags_to_lanes/mod.rs:
```diff
@@ -565,59 +568,191 @@ pub fn tags_to_lanes(
+            }
+            (Some(l), Some(f), None) => (f, num_both_ways, l - f - num_both_ways),
+            (Some(l), None, Some(b)) => (l - b - num_both_ways, num_both_ways, b),
+            (Some(1), None, None) => (0, 1, 0),
```
1 lane becomes a centre lane? seems fishy?

BudgieInWA replied:
It is a bit fishy, but I think it is because describing "how many lanes in each direction does a skinny allyway have" is fishy. I might tag them "lane_markings=no width=5", but we have to represent that in schema somehow.

droogmic commented on rust/osm2lanes/src/transform/tags_to_lanes/mod.rs:
```diff
@@ -565,59 +568,191 @@ pub fn tags_to_lanes(
+        match (tagged_lanes, tagged_forward, tagged_backward) {
+            (_, Some(f), Some(b)) => {
+                if let Some(num_lanes) = tagged_lanes {
+                    if num_lanes != f + b + num_both_ways {
+                        warnings.push(RoadMsg::Ambiguous {
+                            description: None,
+                            tags: Some(tags.subset(&[
+                                "lanes",
+                                "lanes:forward",
+                                "lanes:both_ways",
+                                "lanes:backward",
+                            ])),
+                        });
+                    }
```
Can num_both_lanes could come from center_turn_lanes? Add to warning.

droogmic commented on rust/osm2lanes/src/transform/tags_to_lanes/mod.rs:
```diff
@@ -565,59 +568,191 @@ pub fn tags_to_lanes(
+/// Validates `lanes[:{forward,both_ways,backward}]=*` to determine the precise answer.
+/// Uses `lanes:{bus,psv}[:{forward,backward}]=*`, `busway[:{left,both,right}]=*` and
+/// `centre_turn_lane=yes` to make assumptions in the absence of precise tagging.
+fn driving_lane_directions(
+    tags: &Tags,
+    locale: &Locale,
+    oneway: Oneway,
+    warnings: &mut RoadWarnings,
+) -> (usize, usize, usize) {
```
can we test this in isolation, it is quite complex now :)

BudgieInWA replied:
I agree! (if this function ends up keeping all of its responsibilities.)

Does that mean a set of tests in tests.yaml, or rust tests?

dabreegster replied:
Rust tests I'd imagine, since the input and output are quite specialized here. The unit tests can be in this file, since it's private

droogmic reviewed (changes requested):
Looking good so far, surprisingly complex :)

BudgieInWA commented:
I am finding more and more out about `Tag`s library and the parsing techniques and good `RoadMsg` construction, still a little way to go. But, there is a fundamental problem:

### The problem with locking in a final guess on the lanes distribution at the start

We cannot assume or guess missing parts of the `lanes:{forward,backward}=` scheme without parsing everything else. See how 28300bc adds tons of code parsing other schemes. Definitely smelly.

We need a way to narrow down our guess as we go. Random thoughts:

- Maybe we use `Infer` like values too keep track of guesses.
- Maybe in guessing mode, start with the assumption that there is one normal lane in each direction, then *add lanes* instead of *mutating existing lanes*. Then at the end we can do a round of assuming (e.g. distribute remaining lanes).
- Maybe each tagging scheme has its own `struct` type using `Infer`ed values, which remains internally consistent, to help in the code which validates each tag schema, checks for conflicts, and wants to jump between them in order to do these complicated guesses. Then the validated and consistent schemes are used to build the road. Maybe having code that defines the test and guesses for each requirement between schemes is *exactly the right amount* of code (e.g.  `lanes:bus:forward=2` requires at least 2 forward lanes, or increases the guesses forward lanes by 2), but it still feels a bit like double handling (the whole point of the Builder pattern is to keep track of the jumping between different inputs, right?)
- It ties into validating: `psv:lanes=designated|` requires `lanes=2` to parse properly, so it implies lanes=2 (and is an error if lanes!=2)
- It feels a bit like Logic Programming where different tags are facts that work to collapse the possibility space.

droogmic commented:
> * with the assumption that there is one normal lane in each direction, then _add lanes_ instead of _mutating existing lanes_. Then at

I agree with everything you say, but I also don't really know the best way forward.

> but it still feels a bit like double handling (the whole point of the Builder pattern is to keep track of the jumping between different inputs, right?)

correct

> Maybe in guessing mode, start with the assumption that there is one normal lane in each direction, then add lanes instead of mutating existing lanes. 

I agree, that we should maybe try this instead. The downside is that all the other logic needs to account for this new system, so it will be a painful refactor :(.

A simplified version of the logic you have now should probably still exist, to populate a top level `Infer` field in roadbuilder that may get invalidated by clearer tags later (like your bus example)?

In any event, it sounds like you understand the problem. Now you just need to decide how hard you make the solution for yourself ;)

droogmic commented:
I added you to the project, so hopefully I dont need to approve the CI anymore

BudgieInWA commented:
> In any event, it sounds like you understand the problem. Now you just need to decide how hard you make the solution for yourself ;)

This was just the right thing to hear :) There are a couple of hacky things that I could try, but I don't see them coming together, so I decided to make the solution very hard for myself :D 

> A simplified version of the logic you have now should probably still exist, to populate a top level `Infer` field in roadbuilder that may get invalidated by clearer tags later (like your bus example)?

In [c9ca0eb](https://github.com/a-b-street/osm2lanes/pull/111/commits/c9ca0ebfee220521b2dd72e463dfee37ed3f4907) I started breaking out schema parsing from road building, so that the lanes schema can have access to the already parsed bus lanes. It looks like a direction that I want to keep exploring... which leads to the "hard" part:

Then [0496070](https://github.com/a-b-street/osm2lanes/pull/111/commits/04960707998117b9973ce319f7a5f78485c7a22f) I started writing the overall structure from scratch, as if I had access to `Hunch` and `Scheme` types, and I tried to plot a path through all the steps of assuming the value of some schema given the road, adding tags into the existing assumptions/requirements, etc.

I think `Hunch` needs breaking into two parts, as the comments hint at, but I am actually very happy with `Scheme` and how it helps be think about the problem:
1. Decide a canonical way to represent the scheme - e.g. all of the `busway` tags - in the context of `oneway` and `oneway:bus` can only boil down to: 'is there at least one lane" in each direction(and possibly if it is the opposite direction to expected, depending on interpretations of `opposite_lane`).
3. write `default(&Road) -> Busway` which looks at those dependant tags on the road
4. write `set_tags(mut self, &Tags)` which sets as much detail as possible (and possible makes other assumptions, even guesses), detects any conflicts with already recorded Implications (or should it just read the dependant schemes from the road again, to detect conflicts with implications)
5. write `guess` separately, with whatever crazy logic that I don't want anywhere near the strict-mode parsing path, etc.
6. write `impl Road { update_lanes_with(&mut self, scheme: Busway) {} }` based solely on the values that come out the the `Hunch`es in a schema.

As we think about it more, we can figure out what ordering requirements we need to enforce between all of the `Scheme` updating, and road building steps. It feels like there are comfortable places to put all of the pieces of complexity that I have come across so far: none of the schemes or the assumptions or guesses are that complicated themselves, and this way of dividing the responsibility's up seems to allow each of those things to be expressed in the right context. E.g. "if `lanes=` is not tagged, assume there is one lane in each travel direction (more on freeways) in addition to those described by all the bus tags" is the straightforward assumption that we want to make to support naked `busway` tags and the like. By deciding that busway is handled first, and implementing the assumption logic separate from the tag parsing logic, that becomes the [appropriately straightforward](https://github.com/a-b-street/osm2lanes/pull/111/files#diff-a8762abba191ab492eb7bcb0d196a929c60681561d371c85308a4c3b040e0c0dR180-R221) `let total_bus_lanes_forward = sum_bus_lanes(); ( Assume(default_forward + total_bus_lanes), ...)`.

I would love for @droogmic and/or @dabreegster to have a read through `road.rs` to
- see how close my attempts have approached your existing solution :P,
- see if the way `tags_to_lanes`, `Road`, `Scheme` and `Hunch` fit together in a way is comprehensible, and if so
- help decide what shape `Hunch` and some other types should have, and
- weigh in on how to approach implementing this idea further.

Next steps are
 - decide viability / approach: 
 	- rebuild: add 'strict mode output' to tests, so we can start from scratch and build up parsing of all schemes as written, then come back later to implement guesses (I am happy to get stuck into this kind of approach over some time)
 	- iterate: combine `Hunch` and `Infer`, change `RoadBuilder` to work with `Scheme`, then implement schemes one by one, including existing guessing (sounds sensible if the architectures are similar enough, and we are happy to sacrifice a few tricky tests to begin with)
 - implement type APIs and rust fiddlyness for Hunch and Scheme
 - modify existing RoadBuilder, or write a new one, to work with `Schemes`
 - rewrite all scheme parsing (!) in the new format (which might not be tooo horrible)

BudgieInWA commented:
I'm having too much fun exploring the rust type system to get anything compiling yet, but that is my next step. I should be able to get `Hunch`, `Theory` and `Scheme` compiling easily enough, to start actually seeing if they are useful.

dabreegster commented on rust/osm2lanes/src/transform/tags_to_lanes/mod.rs:
```diff
@@ -243,7 +250,7 @@ impl RoadBuilder {
         .collect();
         // TODO Fix upstream. https://wiki.openstreetmap.org/wiki/Key:centre_turn_lane
-        if tags.is("lanes:both_ways", "1") || tags.is("centre_turn_lane", "yes") {
+        for _ in 0..num_driving_both {
```
Do we have an example of multiple bidirectional lanes? I'm super curious how that would even work in practice O_o

BudgieInWA replied:
I don't really know. I have seen elsewhere the assumption that there can be at most 1 bothways lane (JOSM Lanes I think). 

The reason for the change is actually that the number of bothway lanes is more specifically what we need to look at here, as unlikely as the viability might be - e.g. it supports `lanes:both_ways=2` just fine, even if it breaks someone else's assumptions.

dabreegster commented on rust/osm2lanes/src/transform/tags_to_lanes/road.rs:
```diff
@@ -0,0 +1,412 @@
+// trait HasProblems { fn iter<Problem>... }
+// type TheoryHistory = Vec<Problem>;
+struct Theory<'a, T> {
+    history: Vec<(Hunch<T>, Vec<&'a Theory<'a, T>>)>,
```
It's almost as if we're producing a full proof of how we arrive at the final answer and all of the assumptions made along the way. Really interesting idea, I've never seen something like this before. Interested to see where this approach goes

BudgieInWA replied:
Yes, it just feels silly to throw any of the history away at any of the points where I have the option to. Iterating over the whole set of problems seems like the first useful reason to have any history around.

This line might be the reason I put in the lifetime. I am trying to keep a reference to the `Theory`/`impl TheoryLike` thing in the history, so we don't copy...

dabreegster reviewed (commented):
Sorry for the delay, I'm catching up on everything now (and feeling a little bit lost, but just because I've not looked at this in a few weeks)

I like the approach of teasing apart the schema for everything and having an explicit step where we combine/reconcile. I think it would also nicely handle the cases of lists like `psv:lanes = yes|yes|designated`, where we have an implied total number of lanes that might disagree with a total from another place.

`Hunch` seems like a natural extension of `Infer`. I'm curious where `Theory` will go; it feels complex to understand initially, but possibly very worth it.

droogmic commented on rust/osm2lanes/src/transform/tags_to_lanes/road.rs:
```diff
@@ -0,0 +1,412 @@
+    No,
+}
+
+struct Lanes<'a> {
+    lanes: Theory<'a, Count>,
+    forward: Theory<'a, Count>,
+    backward: Theory<'a, Count>,
+    bothways: Theory<'a, Count>,
+    // directions: Theory<'a, Vec<LaneDirection>>,
+}
+
+struct Busway<'a> {
+    forward: Theory<'a, Option<LaneDirection>>,
+    backward: Theory<'a, Option<LaneDirection>>,
+}
```
having so many lifetime specifiers usually indicates a code smell in rust FYI, as I don't see why you should need them.

BudgieInWA replied:
Yep. I think something forces me to include it, and rust needs assurance here, but I haven't looked into it yet. Hopefully its not needed in the end!

droogmic commented on rust/osm2lanes/src/transform/tags_to_lanes/road.rs:
```diff
@@ -0,0 +1,412 @@
+/// A collection of schemas. We have a Hunch about each schema because we need at least an option,
+/// so why not track it as a Hunch?
+struct Road {
+    busway: Hunch<Busway>,
+    lanes: Hunch<Lanes>,
```
how does this work? do we make assumptions about how lanes vs busways are arranged?

BudgieInWA replied:
We already do, and we need to because people want to tag just `busway:both=lane` to mean 4 lanes: a default 2 normal lanes with a busway on the outside, which we currently support with test [`way_id: 988980354`](https://github.com/a-b-street/osm2lanes/blob/620a83332cd696e08b41ce89e1d559ce7e1a386f/data/tests.yml#L815). It's why `busway` chooses `forward_inside()` vs `forward_outside()`.

droogmic commented on rust/osm2lanes/src/transform/tags_to_lanes/road.rs:
```diff
@@ -0,0 +1,412 @@
+}
+
+/// A value of different certainties. Hunches combine together into another hunch in predefined ways.
+enum Hunch<T> {
+    // Unknown,
+    Tagged(T),
+    Implied(T),
+    Assumed(T),
+    Guessed(T),
+}
```
I think a question you had was asking how to proceed in integrating this. I would start by adding and using this new type in a limited scope and then slowly refactoring out.

Trying to do it all at once will be difficult to make it pan out.

BudgieInWA replied:
I agree, that's why I was happy to split any error tracking out of `Hunch`, to see what I wanted, which is `downgrade` and `reconcile`.

And thanks for being a persistent voice of reason :)

droogmic reviewed (commented):
Good ideas, but the scope has crept a bit ;)

droogmic commented:
At the moment my main concern is how you get such a big change integrated without needing to freeze all parallel development.

I think `Hunch`, and building lanes up from a base assumption of 1 (instead of converting them), are, if I understand correctly the 2 options that we can start with. I wonder to what extend we can deliver those now and expand to this more complete system over time. Things like `Theory` are probably something that can come later.

The reason I say this is that my current short term goal is to push quickly towards getting _most_ roads working in the countries that abstreet and similar tools usually operate, and I think the current simple schema can get us there without too much of an issue.

BudgieInWA commented:
> Good ideas, but the scope has crept a bit ;)
> I'm catching up on everything now (and feeling a little bit lost)...

I told you I had chosen the hard option to look at :) I will keep this idea/practice going in a branch and get back to the task at hand on this one. Thanks for watching me figure out rust.

I will start with pre-parsing the schemes like c9ca0eb, then build up to guessing `lanes=` last. I think it turns out that the order that we consider each scheme matters, even in my experiment. That might be the heart of it, and we can test it with just the schemes that `lanes=` depends on (anything that hints at specific kinds of lanes).

> At the moment my main concern is how you get such a big change integrated without needing to freeze all parallel development.
> ... Things like `Theory` are probably something that can come later.

Absolutely; I will merge main into here as needed once I go back to a compiling commit and we should only implement any ideas bit by bit (maybe we add what I was using in `Hunch` to `Infer`, maybe we add scheme structs with `::from(tags)`, etc.).

BudgieInWA commented:
@droogmic I started focusing on solving the original problem again :)

This is almost ready to merge. Probably a few rust-isms to tweak.

I plan on refactoring all of the bus lane tag parsing to use the `*Scheme` pattern that I have used here, but would prefer not to block this PR on those. As long as the direction that this PR takes `lanes_to_tags` is the right direction, I'm happy to merge, now that most of the tests are passing.

dabreegster commented on rust/osm2lanes/src/transform/tags_to_lanes/bus.rs:
```diff
@@ -55,84 +70,101 @@ pub(super) fn bus(
+}
+
+impl RoadBuilder {
+    pub fn set_busway_scheme(
```
Yeah, this two-step approach of "figure out what's being described" and then "apply it to the road being built up" is way cleaner to read (and even unit test more granularly, if we ever wanted to)

dabreegster commented on rust/osm2lanes/src/transform/tags_to_lanes/bus.rs:
```diff
@@ -55,84 +70,101 @@ pub(super) fn bus(
+            forward_direction: Infer::Default(None),
+            backward_direction: Infer::Default(None),
+        };
+
```
The tiniest of style nitpicks: just one blank newline? cargo fmt may fix for you

BudgieInWA replied:
I have my IDE setup to run cargo format, but it produces different results to `cargo fmt` on the command line (maybe due to the config specifying nightly?). Now I know to run `cargo fmt` before commiting.

droogmic replied:
there is a rustfmt config file, and a toolchain file, that should correctly configure your rust setup. not sure why your IDE might ignore it.

dabreegster commented on rust/osm2lanes/src/transform/tags_to_lanes/lane.rs:
```diff
@@ -0,0 +1,213 @@
+        let tagged_forward: Option<usize> = tags.get_parsed(&(LANES + "forward"), warnings);
+        let tagged_backward: Option<usize> = tags.get_parsed(&(LANES + "backward"), warnings);
+        let tagged_bothways: Option<usize> = tags.get_parsed(&(LANES + "both_ways"), warnings);
+        let tagged_center_turn_lanes = match tags.is("centre_turn_lane", "yes") {
```
Should we emit a deprecation warning? https://wiki.openstreetmap.org/wiki/Key:centre_turn_lane

dabreegster reviewed (commented):
This is pretty fantastic!

droogmic commented on data/tests.yml:
```diff
@@ -794,6 +795,7 @@
 - description: bus:lanes:backward=designated
   skip_python: true
   skip_kotlin: true
+  rust: false # fails roundtrip (outputs busway:right=opposite_lane tag)
```
Can you paste the full error here?
I don't particularly like taking a step backward in passing tests :P

BudgieInWA replied:
I just get this now, which is different to before. I should be able to look into it later, as `set_busway_scheme` is supposed to work with the checks parsing `LanesScheme`. Perhaps more of those checks between schemes that I had as TODO will illuminate the problem... 

```
running 6 tests
test metric::speed::tests::test_speed ... ok
test locale::tests::test_locale ... ok
test metric::tests::test_speed ... ok
test tag::tests::test_tags ... ok
test test::tests::test_roundtrip ... FAILED
test test::tests::test_from_data ... ok

failures:

---- test::tests::test_roundtrip stdout ----
thread 'test::tests::test_roundtrip' panicked at 'called `Result::unwrap()` on an `Err` value: Msg(Unsupported { description: Some("no backward lanes for busway"), tags: None })', osm2lanes/src/test.rs:480:18
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

droogmic replied:
oh yeah, this is the the whole "we need a backward lane to turn into a busway problem".

can we modify the business logic in the edge case to switch from the outside backward lane to the inside forward lane in this case?

droogmic commented on data/tests.yml:
```diff
@@ -1080,39 +1082,67 @@
-    direction: backward
-    designated: motor_vehicle
-  - type: travel
-    designated: foot
+    - type: travel
+      designated: foot
+    - type: travel
+      direction: forward
+      designated: motor_vehicle
+    - type: travel
+      direction: forward
+      designated: motor_vehicle
+    - type: travel
+      direction: backward
+      designated: motor_vehicle
+    - type: travel
+      designated: foot
```
given the new indenting, refactor to the new test scheme:

```
  road:
    highway: road
    lanes:
    - type: travel
      designated: foot
    - type: travel
      direction: forward
      designated: motor_vehicle
    - type: travel
      direction: forward
      designated: motor_vehicle
    - type: travel
      direction: backward
      designated: motor_vehicle
    - type: travel
      designated: foot
```

droogmic replied:
note the one additional layer of nesting.

it saves me doing it later :)

this applies to the other cases as well

BudgieInWA replied:
I think I have updated all the tests that will produce merge conflicts within the `output` lines.

droogmic commented on rust/osm2lanes/src/tag/mod.rs:
```diff
@@ -90,6 +91,23 @@ impl Tags {
     }
 
+    /// Get the value for the given key and parses it into T.
+    pub fn _get_and_parse<V: FromStr>(&self, key: &str) -> Option<V> {
```
a `pub` function with an underscore prefix?
I think both can be removed and then it will be consistent.

droogmic replied:
(you could use this function int he implementation of the below, otherwise I may suggest just removing it for now, as I don't know when we would use it over the below.

droogmic commented on rust/osm2lanes/src/test.rs:
```diff
@@ -115,7 +115,7 @@ impl std::fmt::Display for TestCase {
     let tests: Vec<TestCase> =
         serde_yaml::from_reader(BufReader::new(File::open("../../data/tests.yml").unwrap()))
-            .expect("invalid json");
+            .expect("invalid yaml in data/tests.yml");
```
oops, thanks :D

droogmic commented on rust/osm2lanes/src/transform/lanes_to_tags.rs:
```diff
@@ -84,6 +84,54 @@ pub fn lanes_to_tags(lanes: &[Lane], locale: &Locale, config: &Config) -> TagsRe
+            )
+        }) {
+            tags.checked_insert("lanes:both_ways", "1")?;
+            // TODO: add LHT support
```
emit an unimplemented warning if you want to leave this as is?

BudgieInWA replied:
Got it.

droogmic commented on rust/osm2lanes/src/transform/tags_to_lanes/bus.rs:
```diff
@@ -19,6 +19,21 @@ impl LaneBuilder {
+
+#[derive(Debug)]
+pub struct BuswayScheme {
+    pub forward_direction: Infer<Option<Direction>>,
```
Without looking at the implementation, this is not super intuitive...

So the forward direction is either forward or backwards,
we can infer this to some level of confidence, including inferring that it is neither forward nor backwards?

I will now read the implementation...

droogmic replied:
I think this might just be an `Infer<bool>`

BudgieInWA replied:
See my [other comment](https://github.com/a-b-street/osm2lanes/pull/111#discussion_r846609663) and new name `forward_side_direction`.

If the busway scheme can't describe `forward_side_direction = Backwards` then `forward: Infer<bool>` is better. What about LHT `busway:left=opposite_lane`?

I haven't tried to understand all the assumptions and usecases for busway, so feel free to dictate or even push improvements.

droogmic commented on rust/osm2lanes/src/transform/tags_to_lanes/bus.rs:
```diff
@@ -55,84 +70,101 @@ pub(super) fn bus(
+                );
+            }
+        }
+        if tags.is(
+            BUSWAY + locale.driving_side.opposite().tag(),
+            "opposite_lane",
+        ) {
+            if tags.is("oneway", "yes") || tags.is("oneway:bus", "yes") {
+                // TODO: does it make sense to have a backward lane on the forward_side????
+                busway.forward_direction = Infer::Direct(Some(Direction::Backward));
```
aha, this is the edge case.
So in australia (LHT) given a oneway `busway:right=opposite_lane`

https://wiki.openstreetmap.org/wiki/Key:busway
> Bus lanes in both directions, on left and right sides of a one-way road: [busway]():[left](https://wiki.openstreetmap.org/wiki/Key:left)=opposite_lane + [busway]():[right](https://wiki.openstreetmap.org/wiki/Key:right)=lane

I would expect a bus lane, as the furthest right lane, going in the backward direction (i.e. contraflow).

I struggle with what `busway.forward_direction = Direction::Backward`  means.

droogmic replied:
I guess you are working around the fact that there is no backward_direction lanes to turn into bus lanes?

BudgieInWA replied:
This is from the existing logic. I never did understand it, perhaps it was working around oneway only having forward lanes?

It is probably right to say `backward_direction = Backward` in this case, but that depends on the interpretation of the scheme. I'd say so, and this isn't covered by tests, so I'll make the change. In the context of a `oneway=yes` road "opposite" must mean "backward": 

> Bus lanes in both directions, on left and right sides of a one-way road:
busway:left=opposite_lane + busway:right=lane [RHT]
https://wiki.openstreetmap.org/wiki/Key:busway#Examples

`busway.forward_direction` is better named "forward side direction", so `busway.forward_side_direction = Direction::Backward` is how we describe what the code was doing: `road.forward_inside_mut().direction = Infer::Direct(Direction::Backward)` (though my impl puts the `Direction::Backward` lane on the outside).

droogmic commented on rust/osm2lanes/src/transform/tags_to_lanes/mod.rs:
```diff
@@ -84,17 +90,17 @@ impl std::convert::From<Oneway> for bool {
 #[derive(Copy, Clone, Debug)]
 pub enum Infer<T> {
     None,
+    Guessed(T),
     Default(T),
```
Maybe add some comments/examples of how Guessed differs from default?

I guess something like:

Guessed: no safe default value exists, so a potentially incorrect value was assumed in the absence of any relevant tags.
Default: A safe default value exists, and was used in the absence of any relevant tags.

droogmic commented on rust/osm2lanes/src/transform/tags_to_lanes/mod.rs:
```diff
@@ -84,17 +90,17 @@ impl std::convert::From<Oneway> for bool {
         match self {
             Self::None => None,
-            Self::Default(v) | Self::Direct(v) => Some(v),
+            Self::Guessed(v) | Self::Default(v) | Self::Calculated(v) | Self::Direct(v) => Some(v),
```
I wonder if there are times when we don't want to use a guessed value (due to it's potentially incorrect non-default nature), but this should be fine for now :)

BudgieInWA replied:
Absolutely. Perhaps a `fn some_confident(self) -> Option<T>`. Ideally guesses taint everything they are used to calculate (with appropriate use of `downgrade` type logic), so that by the time we get to the end, we can decide how to render the result (or when we get to road building), depending on the requirements.

droogmic commented on rust/osm2lanes/src/transform/tags_to_lanes/mod.rs:
```diff
@@ -578,59 +618,7 @@ pub fn tags_to_lanes(
-        1
-    };
-    (num_driving_fwd, num_driving_back)
+impl Tags {
```
empty impl?

droogmic commented on rust/osm2lanes/src/transform/tags_to_lanes/mod.rs:
```diff
@@ -546,15 +569,32 @@ pub fn tags_to_lanes(
+    // let bus_lanes = BusLanes::from(tags, locale, &oneway, &mut warnings);
+    // TEMP: lets use the lanes:bus schema to summarise bus lanes for the lanes scheme.
+    let lanes_bus = LanesBusScheme {
+        forward: busway.forward_direction.map_or(0, |o| if o.is_some() { 1 } else { 0 }),
+        backward: busway.backward_direction.map_or(0, |o| if o.is_some() { 1 } else { 0 }),
```
this construct seems fishy, probably due to comments above?

BudgieInWA replied:
Mm, it is a little. I am using `lanes_bus.backward` to count lanes on the backward side of the road (which `BuswayScheme` will switch direction if it needs to). I suppose it depends (if we even need to support such a thing) how `lanes:bus` would represent the contraflow bus lane on the unexpected side of the road situation.

droogmic reviewed (changes requested):
Good progress, thanks.

BudgieInWA commented:
Now we just need to 

- solve our understanding of the busway scheme, at least so [the test passes](https://github.com/a-b-street/osm2lanes/pull/111#discussion_r846289902), and
- merge `main` in here.

@droogmic, I would be happy for you to do the merge, and arrange my new functions wherever you see fit. But if you prefer, I can do it when I have time during the week (my detours are the reason this turned into such a mammoth PR after all).

droogmic commented:
> Now we just need to
> 
>     * solve our understanding of the busway scheme, at least so [the test passes](https://github.com/a-b-street/osm2lanes/pull/111#discussion_r846289902), and
> 
>     * merge `main` in here.
> 
> 
> @droogmic, I would be happy for you to do the merge, and arrange my new functions wherever you see fit. But if you prefer, I can do it when I have time during the week (my detours are the reason this turned into such a mammoth PR after all).

to confirm, you want me to solve the merge conflict? it is indeed probably easier if I do it, given I caused it :P

BudgieInWA commented:
> to confirm, you want me to solve the merge conflict? it is indeed probably easier if I do it, given I caused it :P

Yes, I would appreciate it.

droogmic commented:
The rust part is merged:
https://github.com/BudgieInWA/osm2lanes/pull/1

Some kotlin and python stuff is still failing.

BudgieInWA commented:
Thanks for merging the module restructure @droogmic, I have just merged the newer changes, so this should merge cleanly (currently :P).

I don't know anything about the kotlin tests, and the report linked here didn't show me what was wrong, so can someone else please look into that? Feel free to push commits directly to this branch (which I think is enabled by the "Allow edits by maintainers" checkbox that I have on).

I would like for this PR to be merged now, if we are happy with the direction it takes us. I am happy to be assigned followup issues (or be pinged in fixes to nitpicky issues so I can learn from them) in the interest of working from the same branch going forward. In future, my PRs will be more focused than this one: The detours and explorations that I did on this PR were valuable to me in learning rust and learning this codebase -- especially with your input -- but that's not the way I like to do PRs in general; and I don't plan on restructuring things out of the blue as I ended up doing in this PR.

dabreegster commented on rust/osm2lanes/src/transform/tags_to_lanes/lane.rs:
```diff
@@ -0,0 +1,276 @@
+    }
+}
+
+// struct TurnLanesScheme {
```
nitpick: Remove large chunks of commented code, or use `#[allow(unused)]`. `TurnDirections` sounds like it should wind up being an enum anyway.

dabreegster commented on rust/osm2lanes/src/transform/tags_to_lanes/mod.rs:
```diff
@@ -615,15 +661,45 @@ pub fn tags_to_lanes(
+    let centre_turn_lanes = CentreTurnLaneScheme::new(tags, oneway, locale, &mut warnings);
+    let lanes = LanesScheme::new(
+        tags,
+        /*highway,*/ oneway,
```
Same here -- bits of commented code checked in can add up quickly and get overwhelming. (I'm going to pick on another project I've worked on, https://github.com/Urban-Analytics/RAMP-UA/blob/c5ecee2ab8eebd2b9cffc41956c7c83195b709d1/coding/initialise/population_initialisation.py#L386)

BudgieInWA replied:
Sensible, I'll do a pass, and get some TODOs too.

dabreegster reviewed (approved):
I'm strongly in favor of merging and iterating with smaller cleanups, instead of having more divergence with #138 build up. No worries on a large PR or roundabout way of getting to this point -- learning Rust is (at least) half the fun! Some pretty neat ideas came out of this too.

I'll glance at the kotlin problem -- I spotted something about JSON parsing in the logs.

dabreegster commented:
Kotlin: I can't follow https://github.com/a-b-street/osm2lanes/blob/main/CONTRIBUTING.md#kotlin locally; gradle is throwing some unhelpful error. The test log shows it succeeded at YAML -> JSON, but then `kotlinx.serialization.json.internal.JsonDecodingException at Test.kt:56`.  Unhelpfully, https://github.com/a-b-street/osm2lanes/blob/main/kotlin/src/test/kotlin/Test.kt doesn't have 56 lines, so it's maybe some test framework file?

droogmic commented on data/tests.yml:
```diff
@@ -846,9 +846,10 @@
     sidewalk: "no"
     shoulder: "no"
+    lanes: "3"
+    lanes:forward: "2"
```
can this not be inferred from `bus:lanes:forward`
what happens when this is removed?

droogmic commented on rust/osm2lanes/src/locale.rs:
```diff
@@ -171,6 +171,15 @@ impl std::str::FromStr for DrivingSide {
+impl ToString for DrivingSide {
+    fn to_string(&self) -> String {
+        match self {
+            Self::Right => String::from("right"),
```
not sure if important,

but I avoided implementing this to avoid confusing right hand drive with right hand travel.

droogmic commented on data/tests.yml:
```diff
@@ -846,9 +846,10 @@
     sidewalk: "no"
     shoulder: "no"
+    lanes: "3"
+    lanes:forward: "2"
```
can this not be inferred from bus:lanes:forward
what happens when this is removed?

BudgieInWA replied:
It could, but `lanes:forward` should be tagged in OSM anyway. Inferring it is being a bit fancy and I'm not trying to solve the entirety of lane tagging in this PR.

droogmic replied:
> but lanes:forward should be tagged in OSM anyway

it _should_ but it often isn't, we shouldn't depend on perfect OSM tagging.
we can leave it in to simplify the implementation for now, but please add a comment describing why this data is redundant so that we can try to remove it later?

droogmic commented on rust/osm2lanes/src/locale.rs:
```diff
@@ -171,6 +171,15 @@ impl std::str::FromStr for DrivingSide {
+impl ToString for DrivingSide {
+    fn to_string(&self) -> String {
+        match self {
+            Self::Right => String::from("right"),
```
not sure if important,

but I avoided implementing this to avoid confusing right hand drive with right hand travel.

BudgieInWA replied:
I am confused about the difference, so maybe it is important! Is right hand drive when the driver gets in the right hand door of the car? I suppose this should be `Into<TagKey>` instead of `ToString` or something?

droogmic commented on rust/osm2lanes/src/tag/mod.rs:
```diff
@@ -11,6 +11,8 @@ pub use key::TagKey;
 mod osm;
 pub use osm::{Highway, HighwayType, Lifecycle, HIGHWAY, LIFECYCLE};
 
+use crate::transform::{RoadWarnings, TagsToLanesMsg};
```
it feels strange including these in this mod.

I would expect tag to eventually become its own crate or similar, with no dependencies on the osm2lanes transformation logic.

BudgieInWA replied:
Agreed, `get_parsed` is a convenience function for parsing a bunch of tags at the same time, and should probably live in `transform` somewhere. It was suggested that it be moved in here, but maybe it should be a function, not a method: `let l: Option<usize> = get_parsed_tag(tags, "lanes", &warnings)`; or a method on `RoadWarnings`: `let l: Option<usize> = warnings.parse_tag(tags, "lanes")`?

droogmic commented on rust/osm2lanes/src/tag/mod.rs:
```diff
@@ -90,6 +92,26 @@ impl Tags {
+        &self,
+        key: &K,
+        warnings: &mut RoadWarnings,
+    ) -> Option<T> {
```
`Result<Option<T>, String>`

where the Err is the value that cannot be parsed?

or

`Result<Option<T>, E>`

where E comes from the underlying FromStr

droogmic replied:
maybe Option<Result<T>> makes more sense?

BudgieInWA replied:
It would be `Result<T, E>` (which would be converted to an `Option` with `ok()` most of the time, because the error is already added to the `warnings`).

droogmic commented on rust/osm2lanes/src/transform/tags_to_lanes/lane.rs:
```diff
@@ -0,0 +1,266 @@
+
+const LANES: TagKey = TagKey::from("lanes");
+#[derive(Debug)]
+pub struct LanesScheme {
```
the name of the mod and this struct are unclear for me

in my PR I named this `Counts`, as it isn't really defining a scheme?

BudgieInWA replied:
I didn't know what to name the mod. The struct is named `LanesScheme` because it represents the `lanes:*` tags (which together make up what I was calling "scheme").

droogmic commented on rust/osm2lanes/src/transform/tags_to_lanes/lane.rs:
```diff
@@ -0,0 +1,266 @@
+    pub lanes: Infer<usize>,
+    pub forward: Infer<usize>,
+    pub backward: Infer<usize>,
+    pub bothways: Infer<usize>,
```
`both_ways`.
Rust and the OSM tag both use snake case.

droogmic commented on rust/osm2lanes/src/transform/tags_to_lanes/lane.rs:
```diff
@@ -0,0 +1,266 @@
+
+const CENTRE_TURN_LANE: TagKey = TagKey::from("centre_turn_lane");
+pub struct CentreTurnLaneScheme {
+    pub present: Infer<bool>,
```
what does "present" mean in this context?

BudgieInWA replied:
It refers to the presence of such a lane: "is a centre turn lane present on this road?".

Is something like `struct CentreTurnLaneScheme(Infer<bool>)` possible? I haven't used that language feature before.

droogmic replied:
yes, that would be the correct construct.

droogmic replied:
https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#using-the-newtype-pattern-to-implement-external-traits-on-external-types

droogmic commented on rust/osm2lanes/src/transform/tags_to_lanes/modes/bus.rs:
```diff
@@ -23,6 +23,21 @@ impl LaneBuilder {
+const BUSWAY: TagKey = TagKey::from("busway");
+
+#[derive(Debug)]
+pub struct BuswayScheme {
+    pub forward_side_direction: Infer<Option<Direction>>,
+    pub backward_side_direction: Infer<Option<Direction>>,
+}
```
second: this would be nice to add this as a standalone intermediate refactor PR, to make reviewing easier.

BudgieInWA replied:
> I don't think I found a case of a Backward in forward_side_direction or Forward in backward_side_direction
nor would I expect it.

Given this, I agree that `BuswayScheme(Infer<(bool, bool)>)` is better.

droogmic reviewed (changes requested):
I can agree to this once you look at what I have done and can justify why this is better.

For example:
How do we extend this to include <https://taginfo.openstreetmap.org/keys/taxi:lanes>?
it feels a lot of the lane handling is hardcoded to bus and only bus?

droogmic commented on rust/osm2lanes/src/transform/tags_to_lanes/modes/bus.rs:
```diff
@@ -225,7 +276,7 @@ fn bus_lanes(
             }
             if let Some(backward) = backward {
                 let backward_access = Access::split(backward).map_err(|a| {
-                    RoadError::from(TagsToLanesMsg::unsupported(
+                    RoadError::Msg(TagsToLanesMsg::unsupported(
```
do you remember why you needed to change this? it works as well, I just don't know which is better :P

BudgieInWA replied:
Something silly from my IDE while I was merging. Turns out that wasn't what was breaking the build. We can probably roll this back and I can turn off that warning somehow.
![from error](https://user-images.githubusercontent.com/3784591/164878618-a70b58e3-bcd3-4e81-8eee-ff2ab0763c5e.png)

droogmic replied:
it's fine, I was just curious

droogmic commented on rust/osm2lanes/src/transform/tags_to_lanes/modes/mod.rs:
```diff
@@ -4,8 +4,7 @@ mod bicycle;
 mod bus;
-pub(super) use bus::bus;
-
+pub(super) use bus::{bus, BuswayScheme, LanesBusScheme};
```
caution, whitespace inconsistency

BudgieInWA commented:
> I can agree to this once you look at what I have done and can justify why this is better.
>
> How do we extend this to include [taginfo.openstreetmap.org/keys/taxi:lanes](https://taginfo.openstreetmap.org/keys/taxi:lanes)?
> it feels a lot of the lane handling is hardcoded to bus and only bus?

My intention is to achieve a couple of things:

1. Separate tag parsing from road building, by introducing `Scheme` structs that represent the valid states described by smalls sets of tags. `BuswayScheme` is a good example: the `busway` tags are a complicated way of describing "is there a bus lane on the forward side, and is there a bus lane on the backward side", so we parse all those tags into `BuswayScheme(Infer<(bool, bool)>)` and never have to look up a `busway` tag value again.
2. Fix the guessing of the total number of lanes, by inverting the order that the tags are looked at.
3. Move towards tackling the ordering of looking at different tags at the top level `tags_to_lanes`. a.k.a. express the direction of dependencies between tags in one place. I have only taken a step in that direction, creating `BuswayScheme`, `CentreTurnLaneScheme`, and `LanesScheme` (which take other `Scheme` arguments to show dependencies), and constructing them in the right order at the top level.

These ideas would be further helped by:

- Implementing a `Scheme` for every tag or set of tags that we parse (even `Oneway` has way more complexity that we don't support yet).
- Bringing the guessing logic out of the `Scheme` constructors, and into road builder. Maybe `Scheme` constructors should not `Infer::Guess` at all, and only return `Direct`, `Calculated` (for internally calculated values, like `lanes=lanes:forward + lanes:backward`) and `Default` values. At the top level, or as a set of `RoadBuilder` methods, follow an ordered procedure for building up the road using the `Scheme`s, but no tags.
   - This would allow us to remove all dependencies from `LanesScheme::new`, (except maybe `oneway`) and figure out the guesses in a context which has access to a `RoadBuilder` with some lanes already added and the parsed `LanesScheme` (like https://github.com/a-b-street/osm2lanes/issues/136#issuecomment-1104149579). That step would treat `Direct` values in `LanesScheme` differently to `Default` values.

droogmic commented:
Ok, but then I think we might be trying to do too much in one PR (which is where the difficulty is coming from).

Either we fix the lane counting, or we fix the scheme parsing. Doing both at the same time is blocking quick progress :)

droogmic commented:
kotlin is probably failing in the tests.yml file, so I suggest bisecting the changes you made there until kotlin is happy?

droogmic commented:
my bad, the button for close is very close to the one with comment :laughing:

droogmic commented on data/tests.yml:
```diff
@@ -1158,68 +1161,105 @@
+  skip_kotlin: true
+  tags:
+    highway: secondary
+    lanes: 4
+    lanes:backward: 1
```
these are interpreted as integers, when they should be strings, that causes the kotlin tests to fail

dabreegster commented:
Thanks for splitting into smaller PRs and pushing them through. Just checking in, was closing this PR intentional, or a side effect of the commit message there? Are any of the ideas in this PR still not merged in by one of the smaller PRs?

droogmic commented:
A side-effect. I removed a lot of stuff from this PR that were not needed to make the tests pass

@BudgieInWA will probably want to go through the final status to reintroduce some of the less critical ideas he had (I removed the Infer::Guessed variant for example), preferably as a TDD when they become needed.

BudgieInWA commented:
> A side-effect. I removed a lot of stuff from this PR that were not needed to make the tests pass
> 
> @BudgieInWA will probably want to go through the final status to reintroduce some of the less critical ideas he had (I removed the Infer::Guessed variant for example), preferably as a TDD when they become needed.

This is a great outcome. Getting these ideas our into the open is probably the appropriate amount of progress for now.
